[HDRP] Timeline animation of physical camera / focus distance mode

pmavridis opened the pull request:
### Purpose of this PR
With this PR, we give the option to read the lens focus distance either from the physical camera or the volume component with a new dropdown.  
Also includes some minor UI tweaks that were discussed with UX and our artists.
![image](https://user-images.githubusercontent.com/15788420/115775998-e4ead200-a3b3-11eb-832a-ee110504a9f7.png)

This change will also allow the focus distance to be animated (when it's read from the camera).
Existing projects should be fully compatible with this change.

---
### Testing status
I have tested this manually by switching the dropdown and making sure that we read the focus distance from the right place (either the volume or the camera). 

We need a graphics test for this, but I do not include one in this PR because @TomasKiniulis is currently moving the DoF tests in a single multi-camera one and we can test this new mode in this new multi-camera test.

FrancescoC-unity commented on com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Camera/HDAdditionalCameraData.cs:
```diff
@@ -37,13 +37,24 @@ public class HDPhysicalCamera
         // Lens
         // Note: focalLength is already defined in the regular camera component
         [SerializeField][Range(kMinAperture, kMaxAperture)] float m_Aperture = 16f;
+        [SerializeField][Min(0f)] float m_FocusDistance = 10f;
```
This is bounded differently vs https://github.com/Unity-Technologies/Graphics/pull/4286/files#diff-9ee958b0bde47c58ac148fcbc5dd458acb9a4e0fa124b2ebc3676cc515eae2a4R90 . I am not sure if there was a reason why the one in the volume was not 0, but maybe worth homogenize based on whatever of the two make sense?

pmavridis replied:
Good catch, in the end I kept the same minimum as the volume.

FrancescoC-unity commented on com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Camera/HDAdditionalCameraData.cs:
```diff
@@ -37,13 +37,24 @@ public class HDPhysicalCamera
 
+
+        /// <summary>
+        /// The focus distance of the lens, used in Depth of Field Computations when the DoF volume is configured to read this value from the camera.
```
Computations shouldn't be capitlized here

FrancescoC-unity reviewed (approved):
LGTM, only very minor comments

JordanL8 commented on com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Components/DepthOfField.cs:
```diff
@@ -47,6 +49,22 @@ public enum DepthOfFieldResolution : int
     }
 
+    /// <summary>
+    /// The source of the focus distance used in the Depth of Field Computations
```
Options for the source of the focus distance HDRP uses in the depth of field calculations.

JordanL8 commented on com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Components/DepthOfField.cs:
```diff
@@ -47,6 +49,22 @@ public enum DepthOfFieldResolution : int
+    public enum FocusDistanceMode
+    {
+        /// <summary>
+        /// The focus distance is set on the volume component
```
Uses the focus distance from the Volume component.

JordanL8 commented on com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Components/DepthOfField.cs:
```diff
@@ -47,6 +49,22 @@ public enum DepthOfFieldResolution : int
+        Volume,
+
+        /// <summary>
+        /// The focus distance is set on the physical camera properties
```
Uses the focus distance from the physical camera.

JordanL8 commented on com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Components/DepthOfField.cs:
```diff
@@ -66,11 +84,17 @@ public sealed class DepthOfField : VolumeComponentWithQuality, IPostProcessCompo
 
         /// <summary>
-        /// Sets the distance to the focus point from the Camera.
+        /// Sets the distance to the focus plane from the Camera.
```
The distance to the focus plane from the Camera.

The API docs guidance says to avoids gets/sets - https://confluence.unity3d.com/pages/viewpage.action?spaceKey=DOCS&title=API+docs+content+and+formatting#heading-Propertiesandfields

JordanL8 commented on com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Components/DepthOfField.cs:
```diff
@@ -66,11 +84,17 @@ public sealed class DepthOfField : VolumeComponentWithQuality, IPostProcessCompo
         public MinFloatParameter focusDistance = new MinFloatParameter(10f, 0.1f);
 
+        /// <summary>
+        /// Controls whether the focus distance is read from the physical camera or the volume itself
```
Specifies where to read the focus distance from.

JordanL8 commented on com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Components/DepthOfField.cs:
```diff
@@ -66,11 +84,17 @@ public sealed class DepthOfField : VolumeComponentWithQuality, IPostProcessCompo
+        /// <summary>
+        /// Controls whether the focus distance is read from the physical camera or the volume itself
+        /// </summary>
+        [Tooltip("Controls whether the focus distance is read from the physical camera or the volume itself.")]
```
Specifies where to read the focus distance from.

JordanL8 commented on com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Camera/HDAdditionalCameraData.cs:
```diff
@@ -37,13 +37,24 @@ public class HDPhysicalCamera
 
+
+        /// <summary>
+        /// The focus distance of the lens, used in Depth of Field computations when the DoF volume is configured to read this value from the camera.
```
The focus distance of the lens. The Depth of Field Volume override uses this value if you set `focusDistanceMode` to `FocusDistanceMode.Camera`.

JordanL8 commented on com.unity.render-pipelines.high-definition/Documentation~/Post-Processing-Depth-of-Field.md:
```diff
@@ -20,8 +20,9 @@ Depth Of Field includes [additional properties](More-Options.md) that you must m
-| **Focus Mode**     | Use the drop-down to select the mode that HDRP uses to set the focus for the depth of field effect.<br />&#8226; **Off**: Select this option to disable depth of field.<br />&#8226; **Use Physical Camera**: Select this option to use the physical [Camera](HDRP-Camera.md) to set focusing properties for the depth of field effect. For information on what Camera properties affect ...
-| **Focus Distance** | Set the distance to the focus point from the Camera.<br />This property only appears when you select **Use Physical Camera** from the **Focus Mode** drop-down. |
+| **Focus Mode**     | Use the drop-down to select the mode that HDRP uses to set the focus for the depth of field effect.<br />&#8226; **Off**: Select this option to disable depth of field.<br />&#8226; **Physical Camera**: Select this option to use the physical [Camera](HDRP-Camera.md) to set focusing properties for the depth of field effect. For information on what Camera properties affect dep ...
+| **Focus Distance Mode** | Use the drop-down to select where the focus distance is specified.<br /> &#8226; **Volume**: The focus distance is read from the volume. <br /> &#8226; **Camera**: The focus distance is read from the physical camera. <br /> This property only appears when you select **Physical Camera** from the **Focus Mode** drop-down. |
```
```Use the drop-down to select where the focus distance is specified.<br /> &#8226; **Volume**: Reads the focus distance from the Volume. <br /> &#8226; **Camera**: Reads the focus distance from the physical camera. <br /> This property only appears when you select **Physical Camera** from the **Focus Mode** drop-down.```

sebastienlagarde commented:
@pmavridis  you need to merge master on this one, thanks

iM0ve reviewed (commented):
**Conclusion:** The additional property on the camera is present and usable. However I was not able to animate it.

**Whats tested:**
* Upgrading Amalienborg project that uses Physical cameras
* New property Focus Distance on the camera exists, only has effect when Camera mode is chosen in the DOF override.
* Already existing Focus Distance property found on the DOF override only works when the mode is set to Volume
* Animating Focus Distance of the camera
* Using other Physical Camera properties to see that they are still working

**Conserns** 
Can not animate Focus Distance property using timeline. In the video you can see that I am able to animate FOV, it becomes red when changed and adds a keyframe. Focus Distance however does not behave in the same way, as a matter of fact, none of the Physical Camera properties can be animated with timeline.


https://user-images.githubusercontent.com/36502659/116886443-c3021280-ac31-11eb-85dd-f79d0215d393.mp4


**Project upgrade:**
No issues found here

Master
![Master](https://user-images.githubusercontent.com/36502659/116886520-da410000-ac31-11eb-931e-059390a460bd.jpg)

PR
![Unity_sg3HWuJ5cm](https://user-images.githubusercontent.com/36502659/116886537-df9e4a80-ac31-11eb-849a-0f39bdd06665.jpg)

pmavridis commented:
Regarding Timeline animations, it looks that animation is not working because the phyiscal camera fields are stored in a separate sub-class (HDPhysicalCamera) and not directly in the HDAdditionalCameraData component.
After storing the fields directly in HDAdditionalCameraData aniimation was working in my test, but I need to investigate if we can do that without breaking old projects...

pmavridis commented:
After discussing the animation issue with @RSlysz, we have found that a solution is to convert the class to a struct.
My test shows that this does not require any explicit migration step:
- I changed some physical camera values in an HDRP project before this update.
- Updated HDRP to the version of this PR
- Opened the project and physical camera had the values I had set.

Also animation with Timeline seems to work now:
![image](https://user-images.githubusercontent.com/15788420/117852234-67fe9a00-b287-11eb-84db-afbe7a098216.png)
(it does not work for aperture and curvature because they use some custom UI drawing. We can address those later)
 
[Adding @RSlysz in the reviewers]

RSlysz commented on com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Camera/HDAdditionalCameraData.cs:
```diff
@@ -121,11 +131,30 @@ public void CopyTo(HDPhysicalCamera c)
             c.iso = iso;
             c.shutterSpeed = shutterSpeed;
             c.aperture = aperture;
+            c.focusDistance = focusDistance;
             c.bladeCount = bladeCount;
             c.curvature = curvature;
             c.barrelClipping = barrelClipping;
             c.anamorphism = anamorphism;
         }
```
As it is a struct now and as we copy all field, this method is obsolete.
Basically if assign the struct to another one, it will copy everything inside.

pmavridis replied:
Good point. However this is a public API (and a "blame" shows it has been arround for ~2 years), so I guess it makes sense to keep it for backwards compatibility, right?

RSlysz replied:
Right. But perhaps we can start a deprecation plan like putting it as obsolete (warning, explaining why) then later putting it as obsolete (error) then later removing it. (Can be done later)

@sebastienlagarde do we have any file to keep track of deprecated functionality? And do we have any official deprecation plan? Should we invest little time on having a clear deprecation plan in packages in the next roadmap?

pmavridis replied:
Ok, I marked the method as obsolete, so no need to keep track of this. When we say we will do somethign later, we never do it in the end...

RSlysz commented on com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.PostProcess.cs:
```diff
@@ -292,7 +291,7 @@ void BeginPostProcessFrame(CommandBuffer cmd, HDCamera camera, HDRenderPipeline
             // Grab physical camera settings or a default instance if it's null (should only happen
             // in rare occasions due to how HDAdditionalCameraData is added to the camera)
-            m_PhysicalCamera = camera.physicalParameters ?? m_DefaultPhysicalCamera;
+            m_PhysicalCamera = camera.physicalParameters;
```
just to be 100% sure:
We don't modify it right ?

Before it was shared value by reference. Now it is just a copy.

pmavridis replied:
Yes we don't modify it, it's a copy. 
I kept it like this, in case a post-process pass was modifying the physicalParameters.

RSlysz reviewed (approved):
Seams good. Raised some question just to be sure.

iM0ve reviewed (approved):
With the latest changes I rechecked a few things:
* Physical camera properties still function
* After changes are made to said properties, they persist after editor restart
* Rechecked some DOF override properties, including Physically Based DOF and Quality levels for DOF
* Checked that in the Exposure override, Physical camera property can still be used
* Most importantly, successfully animated the new parameter focus distance on the camera

No more issues on my side

https://user-images.githubusercontent.com/36502659/117966096-34c21680-b32c-11eb-8b67-20e8f993c16b.mp4
